reframe/core/launchers/mpi.py

```diff
@@ -22,6 +22,34 @@ def command(self, job):
         return ['ibrun']
 
 
+@register_launcher('upcrun')
+class UpcrunLauncher(JobLauncher):
+    '''Launcher for UPC applications.'''
+
+    def command(self, job):
+        cmd = ['upcrun']
+        if job.num_tasks:
+            if job.num_tasks_per_node:
+                num_nodes = job.num_tasks // job.num_tasks_per_node
+                cmd += ['-N', str(num_nodes)]
+            cmd += ['-n', str(job.num_tasks)]
+        return cmd
+
+
+@register_launcher('upcxx-run')
+class UpcxxrunLauncher(JobLauncher):
```

Comment:
Why not derive from `UpcrunLauncher` and just change the basic command?

Reply:
good question, I see arguments each way:

 - UPC and UPC++ are similar-but-distinct UPC is a language, an extension of C, while UPC++ is a C++ library. I suspect that the compilers and launchers for each share a lot of code underneath, but they're logically distinct
 - on the other hand, the launchers have (as far as I can tell) the same syntax, so from an implementation perspective they could be considered as belonging to a common class

I can be happy with either implementation, which do you prefer?

Reply:
Thinking a bit more about it, I believe it's better to leave it as is. Other launchers, too, may end up using the same arguments, but it doesn't mean that they have to derive from a common ancestor, since they're conceptually distinct.